src/cnaas_nms/tools/security.py

```diff
@@ -14,10 +28,124 @@ def jwt_required(fn):
+        algorithm = unverified_header.get('alg')
+        try:
+            decoded_token = jwt.decode(token_string, self.keys, algorithms=algorithm, audience=auth_settings.OIDC_CLIENT_ID)
+        except exceptions.ExpiredSignatureError as e:
+            raise ExpiredSignatureError(e)
+        except exceptions.JWTClaimsError as e:
+            try:
+                # with this exception, the call was too fast or the computer too slow
+                logger.info('Checked token too fast (or computer is too slow), retry')
+                sleep(1)
+                decoded_token = jwt.decode(token_string, self.keys, algorithms=algorithm, audience=auth_settings.OIDC_CLIENT_ID)
+            except exceptions.ExpiredSignatureError as e:
+                raise ExpiredSignatureError(e)
+            except exceptions.JWKError as e:
+                logger.error("Invalid Key")
+                raise InvalidKeyError(e)
+        except exceptions.JWKError:
+            try:
+                # with this exception, we first try to reload the keys
+                logger.info('JWT.decode didnt work. Get the keys and retry.')
+                response = requests.get(url=auth_settings.OIDC_CONF_WELL_KNOWN_URL.split('.well-known')[0] + 'oidc/certs')
+                self.keys = response.json()["keys"]
+                # decode the token again
+                decoded_token = jwt.decode(token_string, self.keys, algorithms=algorithm, audience=auth_settings.OIDC_CLIENT_ID)
+            except exceptions.ExpiredSignatureError as e:
+                raise ExpiredSignatureError(e)
+            except exceptions.JWKError as e:
+                logger.error("Invalid Key")
+                raise InvalidKeyError(e)
+        # make an token object to make it easier to validate 
+        token = {
+            "access_token": token_string,
+            "decoded_token": decoded_token,
+            "token_type": algorithm,
+            "audience": auth_settings.OIDC_CLIENT_ID,
+            "expires_at": decoded_token["exp"]
+
+        }
+        return token
+
+    def validate_token(self, token, scopes, request):
+        """Check if token matches the requested scopes."""
+        # if self.scope_insufficient(token.get('scope'), scopes):
+        #     raise errors.InsufficientScopeError()
+        return token
+
+
+oauth_required.register_token_validator(MyBearerTokenValidator())
+
+
+def get_oauth_identity():
+    """Give back the email address of the OAUTH account
+
+        If JWT is disabled, we return "admin".
+
+        We do an api call to request userinfo. This gives back all the userinfo.
+        We get the right info from there and return this to the user.
+
+        Returns:
+            email(str): Email of the logged in user
+
+    """
+    # if jwt disabled, return admin
+    if not api_settings.JWT_ENABLED:
+        return "admin"
+    # request the userinfo
+    url = auth_settings.OIDC_CONF_WELL_KNOWN_URL.split('.well-known')[0] + 'oidc/userinfo'
```

Comment:
Might not be oidc/userinfo in all cases. This is retrievable by reading the .well-know configuration. so probably something likt this:

`
metadata = requests.get("https://yourauthnprovider.example.com/oidc/.well-known
user_info_endpoint = request.json()["userinfo_endpoint"]


`